File: Harvest/ViewModels/AuthViewModel.swift
import Foundation
import Observation
import Supabase
import Auth

@Observable
final class AuthViewModel {
    var isAuthenticated = false
    var isLoading = true
    var profile: UserProfile?
    var error: String?
    var currentUserId: String?

    private let authService = AuthService()
    private let profileService = ProfileService()
    private let subscriptionService = SubscriptionService()

    var needsOnboarding: Bool {
        guard let profile else { return true }
        return profile.onboardingCompleted != true &&
               (profile.age == nil || profile.gender == nil || (profile.photos?.isEmpty ?? true))
    }

    func checkSession() async {
        isLoading = true
        defer { isLoading = false }

        do {
            if let session = try await authService.getCurrentSession() {
                currentUserId = session.user.id.uuidString
                isAuthenticated = true
                await loadProfile()
            }
        } catch {
            isAuthenticated = false
        }
    }

    func login(email: String, password: String) async {
        isLoading = true
        self.error = nil

        do {
            let session = try await authService.signIn(email: email, password: password)
            currentUserId = session.user.id.uuidString
            isAuthenticated = true
            await loadProfile()
        } catch {
            self.error = error.localizedDescription
        }

        isLoading = false
    }

    func register(email: String, password: String) async {
        isLoading = true
        self.error = nil

        do {
            let user = try await authService.signUp(email: email, password: password)
            currentUserId = user.id.uuidString

            // Create profile
            _ = try await profileService.createProfile(userId: user.id.uuidString, email: email)

            // Initialize subscription
            try? await subscriptionService.initializeUserSubscription(userId: user.id.uuidString)

            isAuthenticated = true
            await loadProfile()
        } catch {
            self.error = error.localizedDescription
        }

        isLoading = false
    }

    func logout() async {
        do {
            try await authService.signOut()
        } catch {
            // Ignore signout errors
        }
        isAuthenticated = false
        currentUserId = nil
        profile = nil
    }

    func loadProfile() async {
        guard let userId = currentUserId else { return }
        do {
            profile = try await profileService.getProfile(userId: userId)
        } catch {
            // Profile may not exist yet
        }
    }

    func listenToAuthChanges() {
        Task {
            for await (event, session) in authService.authStateChanges() {
                switch event {
                case .signedIn:
                    if let user = session?.user {
                        currentUserId = user.id.uuidString
                        isAuthenticated = true
                        await loadProfile()
                    }
                case .signedOut:
                    isAuthenticated = false
                    currentUserId = nil
                    profile = nil
                default:
                    break
                }
            }
        }
    }
}

File: android/app/src/main/java/com/harvestglass/harvest/ui/auth/AuthViewModel.kt
package com.harvestglass.harvest.ui.auth

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.harvestglass.harvest.data.model.UserProfile
import com.harvestglass.harvest.data.service.AuthService
import com.harvestglass.harvest.util.userMessage
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

data class AuthUiState(
    val isLoading: Boolean = true,
    val isAuthenticated: Boolean = false,
    val profile: UserProfile? = null,
    val currentUserId: String? = null,
    val error: String? = null
) {
    /**
     * Mirrors AuthViewModel.swift exactly:
     *   onboardingCompleted != true && (age == nil || gender == nil || photos.isEmpty)
     */
    val needsOnboarding: Boolean
        get() {
            val p = profile ?: return true
            return p.onboardingCompleted != true &&
                (p.age == null || p.gender == null || p.photos.isNullOrEmpty())
        }
}

/** Mirrors Harvest/ViewModels/AuthViewModel.swift. */
@HiltViewModel
class AuthViewModel @Inject constructor(
    private val authService: AuthService
) : ViewModel() {

    private val _state = MutableStateFlow(AuthUiState())
    val state: StateFlow<AuthUiState> = _state.asStateFlow()

    fun checkSession() = viewModelScope.launch {
        _state.update { it.copy(isLoading = true) }
        val userId = runCatching { authService.currentUserId() }.getOrNull()
        if (userId == null) {
            _state.update { it.copy(isLoading = false, isAuthenticated = false) }
            return@launch
        }
        val profile = runCatching { authService.loadProfile(userId) }.getOrNull()
        _state.update {
            it.copy(
                isLoading = false,
                isAuthenticated = true,
                currentUserId = userId,
                profile = profile
            )
        }
    }

    fun login(email: String, password: String) = viewModelScope.launch {
        _state.update { it.copy(isLoading = true, error = null) }
        runCatching { authService.signIn(email, password) }
            .onSuccess { userId ->
                val profile = runCatching { authService.loadProfile(userId) }.getOrNull()
                _state.update {
                    it.copy(
                        isLoading = false,
                        isAuthenticated = true,
                        currentUserId = userId,
                        profile = profile
                    )
                }
            }
            .onFailure { e -> _state.update { it.copy(isLoading = false, error = e.userMessage()) } }
    }

    fun register(email: String, password: String) = viewModelScope.launch {
        _state.update { it.copy(isLoading = true, error = null) }
        runCatching { authService.signUp(email, password) }
            .onSuccess { userId ->
                // iOS also creates the profile row and initialises the
                // subscription here; both port with their features in P2.
                _state.update {
                    it.copy(
                        isLoading = false,
                        isAuthenticated = userId != null,
                        currentUserId = userId
                    )
                }
            }
            .onFailure { e -> _state.update { it.copy(isLoading = false, error = e.userMessage()) } }
    }

    fun logout() = viewModelScope.launch {
        runCatching { authService.signOut() }
        _state.value = AuthUiState(isLoading = false, isAuthenticated = false)
    }
}
